lark/visitors.py

```diff
@@ -402,46 +401,47 @@ def inner(cls, tree):
 
 # Decorators
 
-def _apply_decorator(obj, decorator, **kwargs):
+def _apply_v_args(obj, visit_wrapper):
     try:
-        _apply = obj._apply_decorator
+        _apply = obj._apply_v_args
     except AttributeError:
-        return decorator(obj, **kwargs)
+        return _VArgsWrapper(obj, visit_wrapper)
     else:
-        return _apply(decorator, **kwargs)
+        return _apply(visit_wrapper)
 
 
-def _inline_args__func(func):
-    @wraps(func)
-    def create_decorator(_f, with_self):
-        if with_self:
-            def f(self, children):
-                return _f(self, *children)
-        else:
-            def f(self, children):
-                return _f(*children)
-        return f
-
-    return smart_decorator(func, create_decorator)
+class _VArgsWrapper:
+    """
+    A wrapper around a Callable. It delegates `__call__` to the Callable.
+    If the Callable has a `__get__`, that is also delegate and the resulting function is wrapped.
+    Otherwise, we use the original function mirroring the behaviour without a __get__.
+    We also have the visit_wrapper attribute to be used by Transformers.
+    """
+    def __init__(self, func: Callable, visit_wrapper: Callable[[Callable, str, list, Any], Any]):
+        if isinstance(func, _VArgsWrapper):
+            func = func.base_func
+        self.base_func = func
+        self.visit_wrapper = visit_wrapper
+        update_wrapper(self, func)
 
+    def __call__(self, *args, **kwargs):
+        return self.base_func(*args, **kwargs)
 
-def _visitor_args_func_dec(func, visit_wrapper=None, static=False):
-    def create_decorator(_f, with_self):
-        if with_self:
-            def f(self, *args, **kwargs):
-                return _f(self, *args, **kwargs)
+    def __get__(self, instance, owner):
+        try:
+            g = self.base_func.__get__
+        except AttributeError:
+            return self
         else:
-            def f(self, *args, **kwargs):
-                return _f(*args, **kwargs)
-        return f
+            return _VArgsWrapper(g(instance, owner), self.visit_wrapper)
 
-    if static:
-        f = wraps(func)(create_decorator(func, False))
-    else:
-        f = smart_decorator(func, create_decorator)
-    f.vargs_applied = True
-    f.visit_wrapper = visit_wrapper
-    return f
+    def __set_name__(self, owner, name):
+        try:
+            f = self.base_func.__set_name__
+        except AttributeError:
+            return
+        else:
+            f(owner, name)
```

Comment:
This isn't covered by any of the tests

Reply:
That is because I don't know when it might be relevant. AFAIK no standard descriptors (except maybe `property`) actually use `__set_name__`. I just added this for completeness. But I will see if I find something.

Reply:
I added a test. Not the most useful of situations, but one that breaks without the above code (and never would have worked with the old system).